src/python/foglamp/device/ingest.py

```diff
@@ -53,62 +50,246 @@ class Ingest(object):
+
+    _max_idle_db_connection_seconds = 180
+    """Database connection timeout"""
+
+    _max_batch_items = 100
+    """Maximum number of rows in a batch of inserts"""
+
+    _min_queue_flush_seconds = 1
+    """Number of seconds to wait for a queue to reach the maximum batch size"""
+
+    _batch_yield_num_items = 25
+    """While creating a batch, yield to other tasks after this taking this many
+    items from the queue"""
+
+    _max_queue_items = 3*_max_batch_items
+    """Maximum number of items in a queue"""
+
+    _max_insert_attempts = 60
+    """Number of times to attempt to insert a batch"""
+
     @classmethod
-    def start(cls):
+    async def start(cls):
         """Starts the server"""
-        cls._write_statistics_loop_task = asyncio.ensure_future(cls._write_statistics_loop())
+        if cls._started:
+            return
+
+        # TODO: Read config
+
+        # Start asyncio tasks
+        cls._write_statistics_task = asyncio.ensure_future(cls._write_statistics())
+
+        cls._insert_readings_tasks = []
+        cls._insert_readings_wait_tasks = []
+        cls._queues = []
+
+        for _ in range(cls._num_queues):
+            cls._queues.append(asyncio.Queue(maxsize=cls._max_queue_items))
+            cls._insert_readings_wait_tasks.append(None)
+            cls._insert_readings_tasks.append(asyncio.ensure_future(cls._insert_readings(_)))
+
+        cls._started = True
 
     @classmethod
     async def stop(cls):
         """Stops the server
 
-        Saves any pending statistics are saved
+        Flushes pending statistics and readings to the database
         """
-        if cls._stop or cls._write_statistics_loop_task is None:
+        if cls._stop or not cls._started:
             return
 
         cls._stop = True
 
-        if cls._sleep_task is not None:
-            cls._sleep_task.cancel()
-            cls._sleep_task = None
+        for _ in cls._insert_readings_wait_tasks:
+            if _ is not None:
+                _.cancel()
 
-        await cls._write_statistics_loop_task
-        cls._write_statistics_loop_task = None
+        for _ in cls._insert_readings_tasks:
+            await _
+
+        cls._started = False
+
+        cls._insert_readings_wait_tasks = None
+        cls._insert_readings_tasks = None
+        cls._queues = None
+
+        # Write statistics
+        if cls._write_statistics_sleep_task is not None:
+            cls._write_statistics_sleep_task.cancel()
+            cls._write_statistics_sleep_task = None
+
+        await cls._write_statistics_task
+        cls._write_statistics_task = None
+
+        cls._stop = False
 
     @classmethod
     def increment_discarded_readings(cls):
         """Increments the number of discarded sensor readings"""
         cls._discarded_readings += 1
 
     @classmethod
-    async def _write_statistics_loop(cls):
+    async def _close_connection(cls, connection: asyncpg.connection.Connection):
```

Comment:
does with context not work?

Reply:
With context requires using a connection pool which is more overhead than is needed.